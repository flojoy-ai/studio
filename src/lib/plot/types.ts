export interface Drawable {
  draw: () => void;
}
