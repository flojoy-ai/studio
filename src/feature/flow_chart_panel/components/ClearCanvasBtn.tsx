import {
  AlertDialog,
  AlertDialogAction,
  AlertDialogCancel,
  AlertDialogContent,
  AlertDialogDescription,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogTitle,
  AlertDialogTrigger,
} from "@src/components/ui/alert-dialog";
import { Button } from "@src/components/ui/button";
import { Eraser } from "lucide-react";

type ClearCanvasBtnProps = {
  clearCanvas: () => void;
};

export const ClearCanvasBtn = ({ clearCanvas }: ClearCanvasBtnProps) => {
  return (
    <AlertDialog>
      <AlertDialogTrigger asChild>
        <Button
          data-testid="clear-canvas-btn"
          className="gap-2"
          variant="ghost"
        >
          <Eraser />
          Clear Canvas
        </Button>
      </AlertDialogTrigger>
      <AlertDialogContent>
        <AlertDialogHeader>
          <AlertDialogTitle>Are you absolutely sure?</AlertDialogTitle>
          <AlertDialogDescription>
            This action cannot be undone. This will remove everything on the
            flowchart.
          </AlertDialogDescription>
        </AlertDialogHeader>
        <AlertDialogFooter>
          <AlertDialogCancel>Cancel</AlertDialogCancel>
          <AlertDialogAction onClick={clearCanvas}>Continue</AlertDialogAction>
        </AlertDialogFooter>
      </AlertDialogContent>
    </AlertDialog>
  );
};
