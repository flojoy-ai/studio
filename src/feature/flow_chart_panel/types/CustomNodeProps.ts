export type ElementsData = {
  id: string;
  label: string;
  func: string;
  type: string;
  running?: boolean;
  failed?:boolean;
  ctrls: {
    [key: string]: {
      functionName: string;
      param: string;
      value: number;
    };
  };
  inputs?: Array<{ name: string; id: string; type:string }>
  selects?: any;
}

export interface CustomNodeProps {
  data: ElementsData;
}