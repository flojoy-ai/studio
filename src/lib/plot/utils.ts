export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
