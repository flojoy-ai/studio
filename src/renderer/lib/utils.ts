import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
