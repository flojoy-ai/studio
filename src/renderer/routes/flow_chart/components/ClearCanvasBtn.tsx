import {
  AlertDialog,
  AlertDialogAction,
  AlertDialogCancel,
  AlertDialogContent,
  AlertDialogDescription,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogTitle,
  AlertDialogTrigger,
} from "@/renderer/components/ui/alert-dialog";
import { Button } from "@/renderer/components/ui/button";
import { Eraser } from "lucide-react";

type ClearCanvasBtnProps = {
  clearCanvas: () => void;
};

export const ClearCanvasBtn = ({ clearCanvas }: ClearCanvasBtnProps) => {
  return (
    <AlertDialog>
      <AlertDialogTrigger asChild>
        <Button
          data-testid="clear-canvas-button"
          className="gap-2"
          variant="ghost"
        >
          <Eraser className="stroke-muted-foreground" size={20} />
          Clear Canvas
        </Button>
      </AlertDialogTrigger>
      <AlertDialogContent>
        <AlertDialogHeader>
          <AlertDialogTitle>Are you absolutely sure?</AlertDialogTitle>
          <AlertDialogDescription>
            This action cannot be undone. This will remove everything on the
            flowchart.
          </AlertDialogDescription>
        </AlertDialogHeader>
        <AlertDialogFooter>
          <AlertDialogCancel>Cancel</AlertDialogCancel>
          <AlertDialogAction
            data-testid="confirm-clear-canvas"
            onClick={clearCanvas}
          >
            Continue
          </AlertDialogAction>
        </AlertDialogFooter>
      </AlertDialogContent>
    </AlertDialog>
  );
};
