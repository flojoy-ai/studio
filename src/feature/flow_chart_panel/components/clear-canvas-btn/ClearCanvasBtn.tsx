import { createStyles } from "@mantine/core";
import { Draft } from "immer";
import { Edge, Node } from "reactflow";
import { ElementsData } from "../../types/CustomNodeProps";

const useStyles = createStyles((theme) => {
  return {
    button: {
      width: "fit-content",
      height: "43px",
      left: "10px",
      top: "110px",
      margin: "10px",
      boxSizing: "border-box",
      backgroundColor: "transparent",
      color: theme.colors.red[8],
      border: `1px solid ${theme.colors.red[8]}`,
      cursor: "pointer",
      zIndex: 100,
    },
  };
});

type ClearCanvasBtnProps = {
  setNodes: (
    update:
      | Node<ElementsData>[]
      | ((draft: Draft<Node<ElementsData>>[]) => void)
  ) => void;
  setEdges: (update: Edge[] | ((draft: Draft<Edge>[]) => void)) => void;
};

export const ClearCanvasBtn = ({ setNodes, setEdges }: ClearCanvasBtnProps) => {
  const { classes } = useStyles();

  const deleteAllNodes = () => {
    setNodes([]);
    setEdges([]);
  };

  return (
    <button
      data-testid="clear-canvas-btn"
      className={classes.button}
      onClick={deleteAllNodes}
    >
      Clear Canvas
    </button>
  );
};
