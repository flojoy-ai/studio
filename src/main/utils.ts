import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
