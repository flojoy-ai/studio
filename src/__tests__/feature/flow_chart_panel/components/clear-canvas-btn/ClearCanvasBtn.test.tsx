import { renderWithTheme } from "@src/__tests__/__utils__/utils";
import { ClearCanvasBtn } from "@src/feature/flow_chart_panel/components/clear-canvas-btn/ClearCanvasBtn";
import { ElementsData } from "@src/feature/flow_chart_panel/types/CustomNodeProps";
import { fireEvent, renderHook } from "@testing-library/react";
import { useAtom } from "jotai";
import { atomWithImmer } from "jotai-immer";
import { Edge, Node } from "reactflow";

const initialNodes: Node<ElementsData>[] = [
  {
    width: 150,
    height: 135,
    id: "1",
    type: "test",
    data: {
      id: "1",
      label: "LINSPACE",
      func: "LINSPACE",
      type: "SIMULATION",
      ctrls: {},
      selected: false,
    },
    position: {
      x: 0,
      y: 0,
    },
    selected: false,
    positionAbsolute: {
      x: 0,
      y: 0,
    },
    dragging: true,
  },
  {
    width: 150,
    height: 135,
    id: "2",
    type: "test",
    data: {
      id: "2",
      label: "LINSPACE",
      func: "LINSPACE",
      type: "SIMULATION",
      ctrls: {},
      selected: false,
    },
    position: {
      x: 100,
      y: 100,
    },
    selected: false,
    positionAbsolute: {
      x: 100,
      y: 100,
    },
    dragging: true,
  },
];

const initialEdges: Edge[] = [
  {
    id: "e1-2",
    type: "straight",
    source: "1",
    target: "2",
    label: "label",
  },
];

const nodesAtom = atomWithImmer(initialNodes);
const edgesAtom = atomWithImmer(initialEdges);

jest.doMock(
  "@src/feature/flow_chart_panel/components/clear-canvas-btn/ClearCanvasBtn",
  () => {
    const ClearCanvasBtnMock = () => {
      const [nodes, setNodes] = useAtom(nodesAtom);
      const [edges, setEdges] = useAtom(edgesAtom);

      return <ClearCanvasBtn setNodes={setNodes} setEdges={setEdges} />;
    };
    return { ClearCanvasBtn: ClearCanvasBtnMock };
  }
);

const ClearCanvasBtnTest =
  require("@src/feature/flow_chart_panel/components/clear-canvas-btn/ClearCanvasBtn").ClearCanvasBtn;

describe("CanvasClearBtn", () => {
  it("renders correctly with default props", () => {
    const { container } = renderWithTheme(
      <ClearCanvasBtn setNodes={jest.fn()} setEdges={jest.fn()} />
    );
    expect(container).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("clears nodes when clicked", () => {
    const { container, getByTestId } = renderWithTheme(<ClearCanvasBtnTest />);

    const [nodes, setNodes] = renderHook(() => useAtom(nodesAtom)).result
      .current;

    expect(nodes).not.toHaveLength(0);

    const [edges, setEdges] = renderHook(() => useAtom(edgesAtom)).result
      .current;
    expect(edges).not.toHaveLength(0);

    const clearButton = getByTestId("clear-canvas-btn");
    fireEvent.click(clearButton);

    const [newNodes, _setNodes] = renderHook(() => useAtom(nodesAtom)).result
      .current;
    const [newEdges, _setEdges] = renderHook(() => useAtom(edgesAtom)).result
      .current;

    expect(newNodes).toHaveLength(0);
    expect(newEdges).toHaveLength(0);
  });
});
