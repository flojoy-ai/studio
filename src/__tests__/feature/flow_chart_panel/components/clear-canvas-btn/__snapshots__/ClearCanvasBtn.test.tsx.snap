// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasClearBtn renders correctly with default props 1`] = `
<div>
  <button
    class="mantine-UnstyledButton-root mantine-Button-root mantine-2gorln"
    data-button="true"
    data-testid="clear-canvas-btn"
    type="button"
  >
    <div
      class="mantine-1wpc1xj mantine-Button-inner"
    >
      <span
        class="mantine-1ryt1ht mantine-Button-label"
      >
        Clear Canvas
      </span>
    </div>
  </button>
  ,
</div>
`;
